### DISCUSSION TITLE
context menu

### DISCUSSION THREAD

**Discussion by shaharzi** [cefsharp/CefSharp#3540] [Q&A]
Hi,

I have a custom context menu that implement copy / cut/ paste functionality, 1. the problem is that clicking on empty area on the page doesn't close the menu, the menu is stuck. 2. can i get the text copied by the copy function in code? tried the clipboard but the copied text is not the one i copied through the context menu browser.FocusedFrame.Copy();

thanks.

**Comment by amaitland** [comment]
By default CEF uses it's own message loop, messages from Chromium don't propagate to the parent form.

You can integrate CEF into your apps message loop or hook the low level hwnd to get the activate message.

https://github.com/cefsharp/CefSharp/wiki/General-Usage#multithreadedmessageloop
https://github.com/cefsharp/CefSharp/blob/cefsharp/89/CefSharp.WinForms.Example/BrowserTabUserControl.cs#L295

Likely in version 90 I'll add a version of ChromeWidgetMessageInterceptor to simplify hooking the chromium hwnd.

As for selected text IContextMenuParams has contextual specific options including the currently selected text.

http://cefsharp.github.io/api/89.0.x/html/T_CefSharp_IContextMenuHandler.htm
http://cefsharp.github.io/api/89.0.x/html/T_CefSharp_IContextMenuParams.htm

**Reply by shaharzi** [reply]
thanks, i did implmetation by the example, but now i have an odd think that when i make a copy and paste it works just in the second time i press copy, any idea?

using CefSharp;
using CefSharp.Wpf;
using GalaSoft.MvvmLight.Command;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using System.Windows.Controls;

namespace MACEF
{
    internal class MACEFMenuHandler : IContextMenuHandler
    {
        public void OnBeforeContextMenu(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, IContextMenuParams parameters, IMenuModel model)
        {
            model.Clear();

            model.AddItem(CefMenuCommand.Print, "Print");
            model.AddItem(CefMenuCommand.Back, "Back");
            model.AddItem(CefMenuCommand.Forward, "Forward");
            model.AddItem(CefMenuCommand.Reload, "Reload");

            if (parameters.SelectionText.Trim() != "" || Clipboard.GetText().Trim() != "")
            {
                model.AddSeparator();
                model.AddItem(CefMenuCommand.Cut, "Cut");
                model.AddItem(CefMenuCommand.Copy, "Copy");
                model.AddItem(CefMenuCommand.Paste, "Paste");
                model.AddItem(CefMenuCommand.SelectAll, "Select all");
            }
        }

        public bool OnContextMenuCommand(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, IContextMenuParams parameters, CefMenuCommand commandId, CefEventFlags eventFlags)
        {
            
            return false;
        }

        public void OnContextMenuDismissed(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame)
        {
            var webBrowser = (ChromiumWebBrowser)chromiumWebBrowser;

            webBrowser.Dispatcher.Invoke(() =>
            {
                webBrowser.ContextMenu = null;
            });
        }

        public bool RunContextMenu(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, IContextMenuParams parameters, IMenuModel model, IRunContextMenuCallback callback)
        {
            var webBrowser = (ChromiumWebBrowser)chromiumWebBrowser;

            //IMenuModel is only valid in the context of this method, so need to read the values before invoking on the UI thread
            var menuItems = GetMenuItems(model).ToList();

            webBrowser.Dispatcher.Invoke(() =>
            {
                var menu = new ContextMenu
                {
                    IsOpen = true
                };

                RoutedEventHandler handler = null;

                handler = (s, e) =>
                {
                    menu.Closed -= handler;

                    //If the callback has been disposed then it's already been executed
                    //so don't call Cancel
                    if (!callback.IsDisposed)
                    {
                        callback.Cancel();
                    }
                };

                menu.Closed += handler;

                foreach (var item in menuItems)
                {
                    if (item.Item2 == CefMenuCommand.NotFound && string.IsNullOrWhiteSpace(item.Item1))
                    {
                        menu.Items.Add(new Separator());
                        continue;
                    }

                    menu.Items.Add(new MenuItem
                    {
                        Header = item.Item1.Replace("&", "_"),
                        IsEnabled = item.Item3,
                        Command = new RelayCommand(() =>
                        {
                            //BUG: CEF currently not executing callbacks correctly so we manually map the commands below
                            //see https://github.com/cefsharp/CefSharp/issues/1767
                            //The following line worked in previous versions, it doesn't now, so custom EXAMPLE below
                            //callback.Continue(item.Item2, CefEventFlags.None);

                            //NOTE: Note all menu item options below have been tested, you can work out the rest
                            switch (item.Item2)
                            {
                                case CefMenuCommand.Back:
                                    {
                                        browser.GoBack();
                                        break;
                                    }
                                case CefMenuCommand.Forward:
                                    {
                                        browser.GoForward();
                                        break;
                                    }
                                case CefMenuCommand.Cut:
                                    {
                                        browser.FocusedFrame.Cut();
                                        break;
                                    }
                                case CefMenuCommand.Copy:
                                    {
                                        browser.FocusedFrame.Copy();
                                        break;
                                    }
                                case CefMenuCommand.Paste:
                                    {
                                        browser.FocusedFrame.Paste();
                                        break;
                                    }
                                case CefMenuCommand.Print:
                                    {
                                        browser.GetHost().Print();
                                        break;
                                    }
                                case CefMenuCommand.ViewSource:
                                    {
                                        browser.FocusedFrame.ViewSource();
                                        break;
                                    }
                                case CefMenuCommand.Undo:
                                    {
                                        browser.FocusedFrame.Undo();
                                        break;
                                    }
                                case CefMenuCommand.StopLoad:
                                    {
                                        browser.StopLoad();
                                        break;
                                    }
                                case CefMenuCommand.SelectAll:
                                    {
                                        browser.FocusedFrame.SelectAll();
                                        break;
                                    }
                                case CefMenuCommand.Redo:
                                    {
                                        browser.FocusedFrame.Redo();
                                        break;
                                    }
                                case CefMenuCommand.Find:
                                    {
                                        browser.GetHost().Find(0, parameters.SelectionText, true, false, false);
                                        break;
                                    }
                                case CefMenuCommand.AddToDictionary:
                                    {
                                        browser.GetHost().AddWordToDictionary(parameters.MisspelledWord);
                                        break;
                                    }
                                case CefMenuCommand.Reload:
                                    {
                                        browser.Reload();
                                        break;
                                    }
                                case CefMenuCommand.ReloadNoCache:
                                    {
                                        browser.Reload(ignoreCache: true);
                                        break;
                                    }
                                case (CefMenuCommand)26501:
                                    {
                                        browser.GetHost().ShowDevTools();
                                        break;
                                    }
                                case (CefMenuCommand)26502:
                                    {
                                        browser.GetHost().CloseDevTools();
                                        break;
                                    }
                            }
                        })
                    });
                }

                webBrowser.ContextMenu = menu;
            });

            return true;
            //return false;
        }

        private static IEnumerable<Tuple<string, CefMenuCommand, bool>> GetMenuItems(IMenuModel model)
        {
            for (var i = 0; i < model.Count; i++)
            {
                var header = model.GetLabelAt(i);
                var commandId = model.GetCommandIdAt(i);
                var isEnabled = model.IsEnabledAt(i);

                yield return new Tuple<string, CefMenuCommand, bool>(header, commandId, isEnabled);
            }
        }
    }
}

**Reply by amaitland** [reply]
Are you using `WinForms` or `WPF`? You should only need a custom context menu under `WPF` for which you shouldn't have any issues dismissing the context menu via clicking on another element as it's a native `WPF` context menu.

**Reply by shaharzi** [reply]
i using WPF, but when i expose the copy item and use it, i need to paste twice on the area. again - i selecting text,right click, hit copy, and then when i want to paste, only in the second time it paste it.

**Reply by amaitland** [reply]
CefSharp is just one of many chromium embedded framework wrappers. You can download and debug the CefSharp source, you'll see that it's just a simple wrapper.
